File: scripts/codex-max/src-x2/build-superlane.py
#!/usr/bin/env python3
"""Build SRC-X2 apply-ready evidence without changing canonical records."""

from __future__ import annotations

import hashlib
import json
import os
import re
import subprocess
import urllib.error
import urllib.parse
import urllib.request
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
OUT = ROOT / "data/rcap-grade-a/codex-max/source-and-candidate/src-x2"
WAVE = ROOT / "data/rcap-grade-a/launch-control/next-waves/SOURCE_RELATIONSHIP_REPAIR_WAVE.json"
FAMILIES = ROOT / "data/rcap-grade-a/codex-5h/cb05-fleet-index/families.json"
STOPPED = ROOT / "data/rcap-grade-a/codex-5h/cb01-route-integration/stopped.json"
URL_RE = re.compile(r"https?://[^\s\"'<>]+")
OFFICIAL = re.compile(r"(^|\.)(gov|us|courts?\.[a-z]{2}\.us|judiciary\.[a-z]{2}\.gov)$", re.I)


def read(path: Path):
    return json.loads(path.read_text())


def digest(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()


def shard(value: str) -> int:
    return int.from_bytes(hashlib.sha256(value.encode()).digest()[:8], "big") % 8


def dump(name: str, value) -> None:
    (OUT / name).write_text(json.dumps(value, indent=2, sort_keys=True) + "\n")


def tracked_files() -> list[str]:
    result = subprocess.run(["git", "ls-files"], cwd=ROOT, check=True, text=True, capture_output=True)
    return result.stdout.splitlines()


def active_claims():
    ledger = read(ROOT / "data/rcap-grade-a/packet-factory-24h/claim-ledger.json")
    return [c for c in ledger.get("claims", []) if str(c.get("status", "ACTIVE")).upper() not in {"RELEASED", "COMPLETE"}]


def phase1(claims):
    selected = [r for r in read(WAVE)["rows"] if 16 <= int(r["rowId"].split("-")[1]) <= 30]
    rows, patches = [], []
    for r in selected:
        path = ROOT / r["heldPath"] if r.get("heldPath") else None
        candidates = []
        if path and path.is_file():
            candidates.append({"path": r["heldPath"], "sha256": digest(path), "byteLength": path.stat().st_size})
        for candidate in r.get("heldCandidates", []):
            matches = list((ROOT / "private").rglob(candidate["fileName"])) if (ROOT / "private").exists() else []
            candidates.extend({"path": str(p.relative_to(ROOT)), "sha256": digest(p), "byteLength": p.stat().st_size} for p in matches)
        owned = any(any(f in json.dumps(c) for f in r.get("affectedFamilies", [])) for c in claims)
        if owned:
            verdict, reason = "DEFERRED_ACTIVE_CLAUDE_OWNER", "An active Claude claim references an affected family; no claim was modified."
        elif not candidates:
            verdict, reason = "STOPPED_MISSING_BYTES", "No exact held bytes were present at the stated path in this checkout."
        elif r["currentSourceIdentity"]["sourceState"] == "CURRENTNESS_UNVERIFIED":
            verdict, reason = "STOPPED_CURRENTNESS", "Held bytes were measured, but repository evidence does not settle publisher currentness."
        elif len(candidates) != 1:
            verdict, reason = "STOPPED_IDENTITY", "The plausible held identities do not resolve to exactly one measured artifact."
        else:
            verdict, reason = "READY_TO_APPLY", "One measured identity is available and no active Claude ownership was found."
        row = {
            "rowId": r["rowId"], "verdict": verdict, "reason": reason,
            "currentSourceIdentity": r["currentSourceIdentity"], "heldPath": r.get("heldPath"),
            "measuredCandidates": candidates, "plausibleIdentities": [r["currentSourceIdentity"], *r.get("heldCandidates", [])],
            "relationshipDimensions": {"identity": r["currentSourceIdentity"]["canonicalArtifactId"], "currentness": r["currentSourceIdentity"]["sourceState"], "scope": "UNSETTLED_UNLESS_EXPLICIT", "language": "UNSPECIFIED", "filingMode": "UNSPECIFIED", "bundleOrComponent": "UNSPECIFIED", "embeddedSection": "NONE_EVIDENCED", "aliases": r.get("aliases", []), "reuseStatus": "UNSETTLED", "families": r.get("affectedFamilies", [])},
        }
        rows.append(row)
        if verdict == "READY_TO_APPLY":
            patches.append({"operation": "UPSERT_EXACT_SOURCE_RELATIONSHIP", "rowId": r["rowId"], "canonicalArtifactId": r["currentSourceIdentity"]["canonicalArtifactId"], "jurisdiction": r["currentSourceIdentity"]["jurisdiction"], "artifact": candidates[0], "affectedFamilies": r.get("affectedFamilies", []), "applyStatus": "APPLY_READY_NOT_APPLIED"})
    return rows, patches


def normalized_url(raw: str) -> str | None:
    raw = raw.rstrip(".,;:)]}\\")
    try:
        p = urllib.parse.urlsplit(raw)
    except ValueError:
        return None
    host = (p.hostname or "").lower()
    if not p.scheme.startswith("http") or not host or not OFFICIAL.search(host) or any(x in raw for x in ["private/", " ", "|", ",http"]):
        return None
    return urllib.parse.urlunsplit((p.scheme.lower(), p.netloc.lower(), p.path or "/", p.query, ""))


def phase2(files):
    family_ids = [f["familyId"] for f in read(FAMILIES)["families"]]
    support = defaultdict(set)
    receipt_text = ""
    for rel in files:
        low = rel.lower()
        if rel.startswith("data/rcap-grade-a/codex-max/source-and-candidate/src-x2/") or "/candidates/" in low or "candidate" in Path(rel).name.lower():
            continue
        try:
            body = (ROOT / rel).read_text(errors="ignore")
        except (OSError, UnicodeError):
            continue
        if "receipt" in low or "acquisition" in low:
            receipt_text += body
        for raw in URL_RE.findall(body):
            url = normalized_url(raw)
            if url:
                support[url].add(rel)
    owned = [(u, sorted(paths)) for u, paths in support.items() if len(paths) >= 2 and shard(u) == 1 and u not in receipt_text]
    urls, receipts = [], []
    temp = Path(os.environ.get("TMPDIR", "/tmp")) / "legalease-src-x2"
    temp.mkdir(parents=True, exist_ok=True)
    for url, paths in sorted(owned):
        evidence = "\n".join((ROOT / p).read_text(errors="ignore") for p in paths)
        affected = sorted(f for f in family_ids if f in evidence)
        parsed = urllib.parse.urlsplit(url)
        identity = Path(parsed.path).name or parsed.hostname
        rec = {"normalizedUrl": url, "urlSha256": hashlib.sha256(url.encode()).hexdigest(), "shard": 1, "supportingEvidenceFiles": paths, "expectedSourceIdentity": identity, "identityDerivation": "final URL path segment; repository evidence must settle canonical identity before application", "affectedFamilies": affected, "status": "ACQUISITION_BLOCKED", "request": {}}
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "LegalEase-source-verifier/1.0"})
            with urllib.request.urlopen(req, timeout=20) as response:
                body = response.read(30 * 1024 * 1024 + 1)
                ctype = response.headers.get_content_type()
                final = response.geturl()
                if len(body) > 30 * 1024 * 1024:
                    raise ValueError("response exceeds 30 MiB safety limit")
                if ctype == "text/html" and re.search(br"login|sign in|not found|access denied", body[:100000], re.I):
                    raise ValueError("refused login or HTML error body")
                target = temp / hashlib.sha256(url.encode()).hexdigest()
                target.write_bytes(body)
                measured = digest(target)
                rec["request"] = {"httpStatus": response.status, "redirects": final != url, "finalUrl": final, "contentType": ctype, "byteLength": len(body), "sha256": measured, "hashRecomputedFromTemporaryBytes": True}
                rec["status"] = "ACQUISITION_READY"
                receipts.append(rec.copy())
        except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError, ValueError, OSError) as error:
            rec["request"] = {"errorType": type(error).__name__, "error": str(error)[:300]}
        urls.append(rec)
    return urls, receipts


def phase3(claims):
    examined = []
    for f in read(FAMILIES)["families"]:
        if shard(f["familyId"]) != 1 or f.get("claudeOwned"):
            continue
        active = any(f["familyId"] in json.dumps(c) for c in claims)
        controlling = f.get("sourceBound") and f.get("routeMappingStatus") == "BOUND"
        legal_exact = not (f.get("finalBlocker") or {}).get("type") == "LEGAL"
        eligible = not active and controlling and legal_exact and f.get("artifactStatus") != "RENDERED"
        status = "ELIGIBLE" if eligible else "STOPPED"
        reasons = []
        if active: reasons.append("ACTIVE_CLAUDE_OWNER")
        if not controlling: reasons.append("SOURCE_OR_ROUTE_RELATIONSHIP_NOT_EXACT")
        if not legal_exact: reasons.append("CONTROLLING_LEGAL_INPUT_OPEN")
        examined.append({"familyId": f["familyId"], "status": status, "stops": reasons, "implementationStrategyMeasured": f.get("implementationStrategy"), "candidateStatus": "CANDIDATE_BINARY_PROMOTION_PENDING" if eligible else "NOT_BUILT"})
    return examined


def phase4():
    rows = read(STOPPED)["rows"]
    # CB01 calls this key `id`; it is the routeKey whose hash owns a stopped row.
    owned = [r for r in rows if (r.get("routeKey") or r.get("id")) and shard(r.get("routeKey") or r["id"]) == 1]
    return [{"rowId": r["rowId"], "routeKey": r.get("routeKey") or r["id"], "status": "STOPPED_LEGAL_OR_FAMILY_IDENTITY", "reason": r["reason"], "currentResolution": r.get("resolution"), "remeasurement": "CB01 stop remains present at current head; no separate controlling decision record was found that safely removes it.", "canonicalRegistryModified": False} for r in owned]


def main():
    OUT.mkdir(parents=True, exist_ok=True)
    claims = active_claims()
    rows, patches = phase1(claims)
    files = tracked_files()
    urls, receipts = phase2(files)
    families = phase3(claims)
    routes = phase4()
    inputs = [WAVE, FAMILIES, STOPPED, ROOT / "data/rcap-grade-a/launch-control/CLAUDE_9H_SHIFT.json", ROOT / "data/rcap-grade-a/packet-factory-24h/claim-ledger.json", ROOT / "data/rcap-grade-a/packet-factory-24h/SOURCE_CONVEYOR_ASSIGNMENTS.json", ROOT / "data/rcap-grade-a/packet-factory-24h/CHECKPOINT.json"]
    dump("collision-guard.json", {"assignment": "SRC-X2", "readAtHead": subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=ROOT, text=True).strip(), "inputDigests": {str(p.relative_to(ROOT)): digest(p) for p in inputs}, "activeClaudeClaimsObserved": len(claims), "policy": "Claude claims were read only; canonical changes are apply-ready payloads only."})
    dump("source-relationship-rows.json", {"assignedRange": ["SRR-016", "SRR-030"], "expected": 15, "attempted": len(rows), "rows": rows})
    dump("source-relationship-apply-payload.json", {"status": "NOT_APPLIED", "count": len(patches), "operations": patches})
    dump("corroborated-urls.json", {"rule": "first eight SHA-256 bytes, unsigned big-endian, modulo 8 equals 1", "count": len(urls), "urls": urls})
    dump("acquisition-ready-receipts.json", {"sourceBodiesCommitted": 0, "count": len(receipts), "receipts": receipts})
    dump("candidate-families.json", {"examined": len(families), "built": sum(x["status"] == "ELIGIBLE" for x in families), "candidatePdfBinariesCommitted": 0, "families": families})
    dump("route-mapping-payload.json", {"owned": len(routes), "readyToApply": 0, "canonicalRegistryModified": False, "rows": routes})
    ready = sum(r["verdict"] == "READY_TO_APPLY" for r in rows)
    state = {"assignment": "SRC-X2", "baseSha": subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=ROOT, text=True).strip(), "srrAssigned": 15, "srrAttempted": len(rows), "sourceRelationshipsReady": ready, "sourceRelationshipStops": len(rows)-ready, "corroboratedUrlsOwned": len(urls), "acquisitionReadyReceipts": len(receipts), "acquisitionBlocks": len(urls)-len(receipts), "unclaimedFamiliesExamined": len(families), "candidatePacketsBuilt": 0, "candidatePacketsStopped": len(families), "routeMappingReady": 0, "routeMappingStops": len(routes), "sourceBodiesCommitted": 0, "candidatePdfBinariesCommitted": 0, "canonicalRegistriesModified": 0, "packetOverlaysModified": 0, "claimsOrQueuesModified": 0, "commercialRoutesOpened": 0, "productionTouched": False}
    dump("state.json", state)
    summary = "\n".join(f"- {k}: {v}" for k, v in state.items())
    (OUT / "progress.md").write_text(f"# SRC-X2 progress\n\nComplete. All phases continued after stops.\n\n{summary}\n")
    (OUT / "report.md").write_text(f"# SRC-X2 report\n\nThis lane produced apply-ready evidence only. It changed no canonical registry, route, claim, queue, overlay, commercial authority, or Production path. Temporary acquisitions are not committed.\n\n## Result\n\n{summary}\n")


if __name__ == "__main__":
    main()

File: scripts/codex-max/src-x5/build-superlane.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { execFileSync } from 'node:child_process';

const root = execFileSync('git', ['rev-parse', '--show-toplevel'], {encoding:'utf8'}).trim();
process.chdir(root);
const outDir = 'data/rcap-grade-a/codex-max/source-and-candidate/src-x5';
fs.mkdirSync(outDir, {recursive:true});
const read = p => JSON.parse(fs.readFileSync(p, 'utf8'));
const write = (name, value) => fs.writeFileSync(path.join(outDir, name), JSON.stringify(value, null, 2)+'\n');
const sha = value => crypto.createHash('sha256').update(value).digest('hex');
const shard = value => Number(BigInt('0x'+sha(value).slice(0,16)) % 8n);
// Measure the repository head outside this lane so rerunning after the local lane
// commit does not make provenance self-referential or change deterministic output.
const head = execFileSync('git',['log','-1','--format=%H','--','.',':(exclude)data/rcap-grade-a/codex-max/source-and-candidate/src-x5/**',':(exclude)scripts/codex-max/src-x5/**'],{encoding:'utf8'}).trim();
const generatedAt = execFileSync('git',['show','-s','--format=%cI',head],{encoding:'utf8'}).trim();
const tracked = execFileSync('git',['ls-files','-z'],{encoding:'buffer',maxBuffer:16*1024*1024}).toString().split('\0').filter(Boolean);
const textFiles = tracked.filter(p => !p.startsWith(outDir+'/') && !/\.(pdf|png|jpe?g|gif|zip|docx?|xlsx?|woff2?|ico)$/i.test(p));
const textCache = new Map();
const text = p => { if (!textCache.has(p)) { try { textCache.set(p,fs.readFileSync(p,'utf8')); } catch { textCache.set(p,''); } } return textCache.get(p); };
const evidenceFor = needle => textFiles.filter(p => text(p).includes(needle));

const shiftPath='data/rcap-grade-a/launch-control/CLAUDE_9H_SHIFT.json';
const ledgerPath='data/rcap-grade-a/packet-factory-24h/claim-ledger.json';
const checkpointPath='data/rcap-grade-a/packet-factory-24h/CHECKPOINT.json';
const familyPath='data/rcap-grade-a/codex-5h/cb05-fleet-index/families.json';
const wavePath='data/rcap-grade-a/launch-control/next-waves/SOURCE_RELATIONSHIP_REPAIR_WAVE.json';
const collision = {
  schemaVersion:'src-x5-collision-guard/v1', assignment:'SRC-X5', measuredAtHead:head,
  evidence:[shiftPath,ledgerPath,checkpointPath,familyPath],
  rules:{claudeClaimsReadOnly:true, canonicalChangesApplyReadyOnly:true, activeClaudePathsWritable:false},
  ownedWritePaths:[outDir+'/**','scripts/codex-max/src-x5/**'], productionTouched:false, commercialRoutesOpened:0
};
write('collision-guard.json',collision);

const wave=read(wavePath);
const assigned=wave.rows.filter(r=>Number(r.rowId.slice(4))>=61&&Number(r.rowId.slice(4))<=75);
if(assigned.length!==15) throw new Error(`expected 15 SRR rows, got ${assigned.length}`);
const sourceRows=assigned.map(r=>{
  const heldPresent=Boolean(r.heldPath && fs.existsSync(r.heldPath));
  const heldSha256=heldPresent?sha(fs.readFileSync(r.heldPath)):null;
  const candidates=(r.heldCandidates||[]).map(c=>{
    const matches=tracked.filter(p=>path.basename(p)===c.fileName || p.endsWith('/'+c.fileName));
    return {...c, locatedPaths:matches, recomputedSha256:matches[0]&&fs.existsSync(matches[0])?sha(fs.readFileSync(matches[0])):null};
  });
  const claims=r.currentClaims||[];
  const activeClaude=claims.length>0;
  let verdict;
  if(activeClaude) verdict='DEFERRED_ACTIVE_CLAUDE_OWNER';
  else if(!heldPresent && !candidates.some(c=>c.locatedPaths.length)) verdict='STOPPED_MISSING_BYTES';
  else if(r.currentSourceIdentity.sourceState==='FAMILY_IDENTITY_AMBIGUOUS') verdict='STOPPED_IDENTITY';
  else if(r.currentSourceIdentity.sourceState==='CURRENTNESS_UNVERIFIED') verdict='STOPPED_CURRENTNESS';
  else if(r.currentSourceIdentity.sourceState==='SOURCE_SCOPE_AND_VERSION_AMBIGUITY') verdict='STOPPED_SCOPE_OR_VARIANT';
  else verdict='STOPPED_FAMILY_MAPPING';
  const id=r.currentSourceIdentity.canonicalArtifactId;
  const identityEvidence=evidenceFor(id).slice(0,50);
  return {rowId:r.rowId, verdict, currentSourceIdentity:r.currentSourceIdentity, held:{path:r.heldPath,present:heldPresent,sha256:heldSha256},
    plausibleIdentities:[{artifactId:id,relationship:'declared canonical identity'},...candidates.map(c=>({artifactId:c.artifactId,fileName:c.fileName,sha256:c.recomputedSha256||c.sha256,relationship:'held candidate'}))],
    dimensions:{identity:id,currentness:r.currentSourceIdentity.sourceState,scopeOrVariant:r.ambiguity,language:'not distinguished by current evidence',filingMode:'not settled by this repair row',bundleOrComponent:r.currentSourceIdentity.sourceState==='BUNDLE_COMPONENT'?'component':'not identified as component',embeddedSection:r.currentSourceIdentity.sourceState==='EMBEDDED_SECTION',aliases:r.aliases||[],reuseStatus:'not authorized',familyRelationship:r.affectedFamilies||[]},
    activeClaudeClaims:claims, evidenceFiles:[wavePath,...identityEvidence], stopReason:verdict==='STOPPED_CURRENTNESS'?'Held bytes were hashed, but repository evidence does not establish publisher currentness.':r.ambiguity,
    canonicalPatchEligible:verdict==='READY_TO_APPLY'};
});
write('source-relationship-rows.json',{schemaVersion:'src-x5-source-relationship-rows/v1',assignment:'SRC-X5',measuredAtHead:head,assignedRange:['SRR-061','SRR-075'],expected:15,attempted:sourceRows.length,rows:sourceRows});
const ready=sourceRows.filter(r=>r.verdict==='READY_TO_APPLY');
write('source-relationship-apply-payload.json',{schemaVersion:'src-x5-source-relationship-apply-payload/v1',applyStatus:'APPLY_READY_PAYLOAD_ONLY',canonicalRegistryModified:false,measuredAtHead:head,count:ready.length,patches:ready.map(r=>({rowId:r.rowId,operation:'bind_exact_source_relationship',artifactId:r.currentSourceIdentity.canonicalArtifactId,sha256:r.held.sha256,families:r.dimensions.familyRelationship}))});

// A corroborated URL must occur in two separate tracked, non-candidate evidence files.
const urlMap=new Map();
const urlRe=/https?:\/\/[^\s"'<>\\)\]]+/g;
for(const p of textFiles.filter(p=>!p.includes('/candidates/'))){ for(const raw of text(p).match(urlRe)||[]){
  let normalized; try { const u=new URL(raw.replace(/[.,;:]+$/,'')); u.hash=''; normalized=u.toString(); } catch { continue; }
  if(!urlMap.has(normalized)) urlMap.set(normalized,new Set()); urlMap.get(normalized).add(p);
}}
const officialHost=h=>/\.gov$/i.test(h)||/\.gov\./i.test(h)||/\.us$/i.test(h)||/^(courts?|judicial)\.[a-z]{2}\.gov$/i.test(h);
const existingReceiptText=textFiles.filter(p=>/receipt|acquisition/i.test(p)).map(p=>text(p)).join('\n');
const qualified=[];
for(const [url,files] of urlMap){ const u=new URL(url); if(files.size<2||!officialHost(u.hostname)||shard(url)!==4||existingReceiptText.includes(url)) continue;
  qualified.push({normalizedUrl:url,urlSha256:sha(url),shardIndex:4,supportingEvidenceFiles:[...files].sort(),expectedSourceIdentity:'requires exact artifact identity confirmation before acquisition',affectedFamilies:[],status:'ACQUISITION_BLOCKED_IDENTITY',block:'No exact unsettled artifact identity can safely be inferred from URL text alone; guessed downloads are forbidden.'});
}
qualified.sort((a,b)=>a.normalizedUrl.localeCompare(b.normalizedUrl));
write('corroborated-urls.json',{schemaVersion:'src-x5-corroborated-urls/v1',measuredAtHead:head,qualification:{minimumDistinctEvidenceFiles:2,shardIndex:4,shardCount:8},ownedCount:qualified.length,urls:qualified});
write('acquisition-ready-receipts.json',{schemaVersion:'src-x5-acquisition-ready-receipts/v1',measuredAtHead:head,sourceBodiesCommitted:0,count:0,receipts:[],blocks:qualified.map(x=>({normalizedUrl:x.normalizedUrl,status:x.status,reason:x.block}))});

const families=read(familyPath).families;
const ownedFamilies=families.filter(f=>shard(f.familyId)===4);
const familyRows=ownedFamilies.map(f=>{
  let status,reason;
  if(f.claudeOwned||f.currentOwnerScope==='pre-existing-active-assignment'){status='DEFERRED_ACTIVE_CLAUDE_OWNER';reason=`Owner ${f.currentOwner||'unknown'} is active; no candidate files written.`;}
  else if(f.finalBlocker?.type==='LEGAL'||f.currentState==='LEGAL_BLOCKED'){status='STOPPED_LEGAL_TREATMENT';reason=f.finalBlocker?.exact||'Controlling legal treatment is not settled.';}
  else if(!f.sourceBound&&f.implementationStrategy!=='custom_pleading'){status='STOPPED_SOURCE_RELATIONSHIP';reason=f.sourceStatus;}
  else if(f.routeMappingStatus!=='BOUND'){status='STOPPED_FAMILY_MAPPING';reason=f.routeMappingStatus;}
  else if(f.artifactStatus==='RENDERED'&&f.completenessStatus==='PASS_COMPLETE'){status='ALREADY_CANONICAL_COMPLETE';reason='Current index reports a rendered complete artifact.';}
  else {status='STOPPED_CONTROLLING_EVIDENCE';reason='The fleet index does not itself prove controlling and exact legal treatment sufficient to author new pleading text.';}
  return {familyId:f.familyId,jurisdiction:f.jurisdiction,shardIndex:4,examined:true,implementationStrategy:f.implementationStrategy,status,reason,evidence:f.evidence||[],candidateDirectory:null,binaryManifest:null};
});
write('candidate-families.json',{schemaVersion:'src-x5-candidate-families/v1',measuredAtHead:head,selectionRule:'unsigned_big_endian(first_8_bytes(SHA-256(familyId))) mod 8 == 4',examined:ownedFamilies.length,built:0,stopped:familyRows.length,candidatePdfBinariesCommitted:0,families:familyRows});

const stopped=read('data/rcap-grade-a/codex-5h/cb01-route-integration/stopped.json').rows;
const routeRows=stopped.filter(r=>shard(r.id)===4).map(r=>({rowId:r.rowId,routeKey:r.id,jurisdiction:r.jurisdiction,status:'STOPPED_REQUIRES_CANONICAL_REPAIR',reason:r.reason,exactStop:r.resolution?.requiredAction||'Current stopped evidence does not settle an apply-ready mapping.',evidence:[r.evidenceSource,'data/rcap-grade-a/codex-5h/cb01-route-integration/stopped.json']}));
write('route-mapping-payload.json',{schemaVersion:'src-x5-route-mapping-payload/v1',measuredAtHead:head,canonicalRouteRegistryModified:false,ownedCount:routeRows.length,readyToApply:0,stops:routeRows.length,rows:routeRows});

const counts={srrAttempted:sourceRows.length,srrReady:ready.length,srrStops:sourceRows.length-ready.length,urlsOwned:qualified.length,receipts:0,acquisitionBlocks:qualified.length,familiesExamined:ownedFamilies.length,candidatesBuilt:0,candidatesStopped:familyRows.length,routeReady:0,routeStops:routeRows.length};
write('state.json',{schemaVersion:'src-x5-state/v1',assignment:'SRC-X5',baseSha:head,status:'COMPLETE_WITH_RECORDED_STOPS',generatedAt,counts,invariants:{sourceBodiesCommitted:0,candidatePdfBinariesCommitted:0,canonicalRegistriesModified:0,packetOverlaysModified:0,claimsOrQueuesModified:0,commercialRoutesOpened:0,productionTouched:false}});
fs.writeFileSync(path.join(outDir,'progress.md'),`# SRC-X5 progress\n\n- Phase 1: attempted all **${counts.srrAttempted}** assigned rows; ${counts.srrReady} apply-ready and ${counts.srrStops} stopped/deferred.\n- Phase 2: derived **${counts.urlsOwned}** owned corroborated URLs; ${counts.receipts} receipts and ${counts.acquisitionBlocks} safe acquisition blocks.\n- Phase 3: examined **${counts.familiesExamined}** shard-owned families; no family passed all controlling-evidence and collision gates.\n- Phase 4: re-measured **${counts.routeStops}** shard-owned stopped route rows; none was safe to map without its named canonical repair.\n`);
fs.writeFileSync(path.join(outDir,'report.md'),`# SRC-X5 report\n\nThis lane produced apply-ready sidecars only. It did not change canonical registries, claims, queues, overlays, commercial authority, source bodies, or PDF binaries. Every stop remains explicit in the JSON evidence.\n\n## Counts\n\n${Object.entries(counts).map(([k,v])=>`- ${k}: ${v}`).join('\n')}\n`);
console.log(JSON.stringify(counts));
